files/upload_sig.cgi

```diff
@@ -0,0 +1,282 @@
+
+    for group in groupmemberships:
+        if group == conf.get('acls', 'push_superadmin_group'):
+            return True
+
+        if re.match(r'c\ds?-{0}.*'.format(group), branchname):
+            print >>sys.stderr, "Matched {} against {}".format(group,
+                                                               branchname)
+            return True
+
+    # requires to be a sig member
+    if not is_sig_member(groupmemberships):
+        return False
+
+    return False
+
+
+def get_memberships(username):
+    os.environ['KRB5_CLIENT_KTNAME'] = conf.get('acls', 'fasjson_keytab')
+    fasjson_url = conf.get('acls', 'fasjson_url')
+    fasjson_full_url = '%s/v1/users/%s/groups/' % (fasjson_url,username)
+    try:
+      gssapi_auth = HTTPSPNEGOAuth(opportunistic_auth=True,mutual_authentication='OPTIONAL')
+      results = requests.get(fasjson_full_url, auth=gssapi_auth) 
+    except:
+        print >>sys.stderr, "Error looking up group memberships. HTTP Error code {}".format(results.status_code)
+        return None
+      
+    if results.status_code >= 400:
+        print >>sys.stderr, "Error looking up group memberships. HTTP Error code {}".format(results.status_code)
+        return None
+
+    json_body = json.loads(results.text)
+    group_list = []
+    for group in json_body['result']:
+      group_list.append(group['groupname'])
+  
+    return group_list
+
+
+def check_form(form, var, exit_on_error=True):
+    ret = form.getvalue(var, None)
+    if ret is None:
+        if not exit_on_error:
+            return None
+        send_error('Required field "%s" is not present.' % var)
+    if isinstance(ret, list):
+        send_error('Multiple values given for "%s". Aborting.' % var)
+    return ret
+
+
+def send_email(pkg, sha1, filename, username, branch=''):
+    text = """A file has been added to the lookaside cache for %(pkg)s:
+
+%(branch)s %(sha1)s  %(filename)s""" % locals()
+    msg = MIMEText(text)
+    sender_name = conf.get('mail', 'sender_name')
+    sender_email = conf.get('mail', 'sender_email')
+    sender = Utils.formataddr((sender_name, sender_email))
+    recipient = conf.get('mail', 'recipient')
+    msg['Subject'] = 'File %s uploaded to lookaside cache, branch %s by %s' % (
+            filename, branch, username)
+    msg['From'] = sender
+    msg['To'] = recipient
+    try:
+        s = smtplib.SMTP(conf.get('mail', 'smtp_server'))
+        s.sendmail(sender, recipient, msg.as_string())
+    except:
+        errstr = 'sending mail for upload of %s failed!' % filename
+        print >> sys.stderr, errstr
+        syslog.syslog(errstr)
+
+
+def main():
+    os.umask(002)
+
+    username = os.environ.get('SSL_CLIENT_S_DN_CN', None)
+
+    assert os.environ['REQUEST_URI'].split('/')[1] == 'sources'
+
+    form = cgi.FieldStorage()
+    name = check_form(form, 'name')
+    branch = check_form(form, 'branch', exit_on_error=False)
+    is_hash_format = branch is None
```

Comment:
It will deal with different structures based on the `branch` parameter, whenever it is present or not - is this an acceptable approach?

Reply:
I think that makes sense. SIG Uploads just wouldn't pass the 'branch' parameter at all